lib/Dist/Zilla/Dist/Builder.pm

```diff
@@ -486,7 +503,12 @@ sub build_archive {
   my $basename = $self->dist_basename;
   my $basedir = path($basename);
 
-  $_->before_archive for @{ $self->plugins_with(-BeforeArchive) };
+  $_->before_archive for $self->plugins_with(-BeforeArchive)->@*;
+
+  for my $builder ($self->plugins_with(-ArchiveBuilder)->@*) {
+    my $file = $builder->build_archive($self->archive_filename, $built_in, $basename, $basedir);
+    return $file if defined $file;
```

Comment:
On further reading, I think it's "you're not gonna need it" to make an ArchiveNamer plugin.  I think it's just going to be complexity that doesn't get us anything.  Do you have concrete use cases in mind?

Instead, I'd let DZ's Builder code come up with a base name on its own (as it already does), then pass that base name to the ArchiveBuilder which can build the archive and return the full name.

Reply:
If you really want to need to customize the archive name, I'd pass a sprintf-ish style format string to specify the template. (see Dist::Zilla::Plugin::NextRelease and Dist::Zilla::Plugin::Run for other examples of templating short strings.)

Reply:
The application initially intended is to set the extension on the archive for alternative formats.  Allowing .zip, for example I think is not desirable tbh for CPAN (although there are some dists that do this), but I also use dzil for stuff not intended for CPAN where this could be useful.  I viewed this approach as less invasive change in my initial implementation, but I agree that computing the extension inside the builder and returning the complete path is better because you could don't have to worry about an ArchiveBuilder+ArchiveNamer competing with another plugin that does ArchiveNamer.  That said, when I re-worked it below to remove the ArchiveNamer, it actually seems pretty non-invasive.

I think the ArchiveNamer might be useful for Perl-dists on non-CPAN type applications, but I don't have a concrete use case for now, and with the iteration below it would probably operate on the archive_basename rather than the full archive_filename anyway.  In the very unlikely situation I run into a need for this I think we can iterate later.